src/cpp/session/modules/rmarkdown/SessionRmdNotebook.cpp

```diff
@@ -172,9 +173,25 @@ bool fixChunkFilename(int, const core::FilePath& path)
 
 void onChunkExecCompleted(const std::string& docId, 
                           const std::string& chunkId,
+                          const std::string& code,
+                          const std::string& label,
                           const std::string& nbCtxId)
 {
-   emitOutputFinished(docId, chunkId, ExecScopeChunk);
+   r::sexp::Protect rProtect;
+   SEXP resultSEXP = R_NilValue;
+   json::Array results;
+
+   r::exec::RFunction func(".rs.executeChunkCallback");
+   func.addParam(label);
+   func.addParam(code);
```

Comment:
At some point we will also need to pass the results into this R function. Those may be nontrivially expensive to collect, so we might consider asking R whether the chunk callback environment exists first.

Reply:
I updated the tools code so that the environment is always created rather than only once it is in use.